### PR TITLE
Inverse requirement mangling and @_preInverseGenerics

### PR DESCRIPTION
**Pull request by Azoy**
This patch implements mangling support for inverse generics as well as an attribute `@_preInverseGenerics` that can be applied to members that will completely ignore inverse requirements entirely.

### CODE REVIEW

**Comment by tbkka** [comment]
@kavon Does this also need to support ~Escapable?

**Comment by Azoy** [comment]
@tbkka it should yeah. I originally had something like `requirement ::= protocol 'Ri' GENERIC-PARAM-INDEX`, but Slava mentioned we could shorten the mangling because there probably won't be a lot of invertible protocols. So I shortened it to `requirement ::= 'Ri' GENERIC-PARAM-INDEX` assuming the copyable protocol and we can add `RI` to mean escapable for example.

**Comment by tbkka** [comment]
Since ~Copyable and ~Escapable generics are landing at the same time, we should try to make sure we support both of them whenever we have changes like this.

File: lib/AST/ASTMangler.cpp [hunk 1]
```diff
@@ -2568,9 +2648,52 @@ void ASTMangler::appendSymbolicReference(SymbolicReferent referent) {
...
+  //     struct X<Y: ~Copyable> {
+  //       func foo() {} // In order for this to have the same mangling as the
+  //                        one in the extension, we need to treat it as if it
+  //                        were in the extension.
+  //     }
+  //
+  //     extension X where Y: ~Copyable {
+  //       func foo() {}
+  //     }
+  //
```
Comment (kavon) [inline lib/AST/ASTMangler.cpp:2674]: 🤔

**Pull request closed without merging**